### PR TITLE
python3Packages.pyyaml: 5.4.1.1 -> 6.0

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
https://github.com/yaml/pyyaml/releases/tag/6.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
This should probably be done as part of a `python-unstable` jobset. This will likely cause alot of regressions

**Comment by mweinelt** [comment]
Yes, I tried this already due to home-assistant wanting it and it was bad.

**Comment by SuperSandro2000** [comment]
I am not sure how we want to do this upgrade. Can python do overloading and we could create a temporary wrapper?

**Comment by mweinelt** [comment]
> I am not sure how we want to do this upgrade.

In due time and carefully. We will not be able to not break stuff in the process unfortunately. Applications will receive overrides and that's it.

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Review by theotherjimmy — APPROVED** [review]
Would love this for packaging pyocd.

**Comment by dotlambda** [comment]
> Would love this for packaging pyocd.

You can just use the older PyYAML, https://github.com/pyocd/pyOCD/commit/aa4ab29cb36140db6c0a9d164fd5d57e6083be6e makes me believe version 6 is not actually required.

**Review by jonringer — APPROVED** [review]
LGTM

```
https://github.com/NixOS/nixpkgs/pull/145807

1 package built:
azure-cli
```

**Comment by trofi** [comment]
Bisect says `python3Packages.pyyaml: 5.4.1.1 -> 6.0` db01e386f4fe725bfd1f733c26e74dc838156233 broke `awscli` in `staging`:

```
$ nix build -f. awscli -L
...
awscli> installing
awscli> Executing pipInstallPhase
awscli> /build/awscli-1.22.14/dist /build/awscli-1.22.14
awscli> Processing ./awscli-1.22.14-py3-none-any.whl
awscli> Requirement already satisfied: colorama<0.4.4,>=0.2.5 in /nix/store/lsby77isz3iaivz4brh4h28rgpydxfg5-python3.9-colorama-0.4.3/lib/python3.9/site-packages (from awscli==1.22.14) (0.4.3)
awscli> Requirement already satisfied: rsa<5,>=3.1.2 in /nix/store/6avnhxkxy1h6h0viw8afpw70bp626rsv-python3.9-rsa-4.8/lib/python3.9/site-packages (from awscli==1.22.14) (4.8)
awscli> Requirement already satisfied: s3transfer<0.6.0,>=0.5.0 in /nix/store/6gqf8v5n55mdd7i3avijccn0xm36qlnl-python3.9-s3transfer-0.5.0/lib/python3.9/site-packages (from awscli==1.22.14) (0.5.0)
awscli> ERROR: Could not find a version that satisfies the requirement PyYAML<5.5,>=3.10 (from awscli) (from versions: none)
awscli> ERROR: No matching distribution found for PyYAML<5.5,>=3.10
```

**Comment by theotherjimmy** [comment]
Seems like we will need a pyyaml6 and pyyaml5?

**Comment by dotlambda** [comment]
@trofi Can you test whether awscli still works with
```diff
diff --git a/pkgs/tools/admin/awscli/default.nix b/pkgs/tools/admin/awscli/default.nix
index 0b9d6b47383..b04c554be84 100644
--- a/pkgs/tools/admin/awscli/default.nix
+++ b/pkgs/tools/admin/awscli/default.nix
@@ -30,7 +30,8 @@ with py.pkgs; buildPythonApplication rec {
   # https://github.com/aws/aws-cli/issues/4837
   postPatch = ''
     substituteInPlace setup.py \
-      --replace "docutils>=0.10,<0.16" "docutils>=0.10"
+      --replace "docutils>=0.10,<0.16" "docutils>=0.10" \
+      --replace "PyYAML>=3.10,<5.5" "PyYAML>=3.10"
   '';

   propagatedBuildInputs = [
```
Otherwise, we'll have to override PyYAML in `packageOverrides`.

**Comment by trofi** [comment]
@dotlambda your patch fixes build for me. The caveat is that I did not runtime-test it. I only noticed `awscli` build failure when checked unrelated changes locally.

**Comment by jonringer** [comment]
According to the changelog, there was a breaking change with `yaml.load()`: https://github.com/yaml/pyyaml/blob/master/CHANGES

We should probably just add it to the packagesOverride

**Comment by mweinelt** [comment]
Mentioned here: https://github.com/NixOS/nixpkgs/pull/145807#issuecomment-969242447

Sorry, I said this should go via python-unstable and explicitly asked if someone was comfortable enough with this going via staging. I wasn't.

**Comment by jonringer** [comment]
I was comfortable doing it through staging-next.

In general, there's a 2 week cadence for staging during non-ZHF periods, which is plenty of time to stabilize

**Comment by jonringer** [comment]
the current staging-next will take another week at this pace. I already have a lot of python-unstable changes that I would like to do for the next staging-next. Which would push this back to 3 weeks before getting into the next python-unstable; or just merge to staging, and do stabilization then.

**Comment by trofi** [comment]
Bisected another failure down to this bump on `staging`: `lirc`:

```
$ nix build -f. -L lirc
...
lirc> PYTHONPATH=../python-pkg/lirc \
lirc> /nix/store/gn4gyzv93izh3lg5iqjb7l04k9r4hxv0-python3-3.9.9-env/bin/python ./data2hwdb ../configs ../configs \
lirc>     > lirc.hwdb
lirc> Traceback (most recent call last):
lirc>   File "/build/lirc-0.10.1/doc/./data2hwdb", line 112, in <module>
lirc>     main()
lirc>   File "/build/lirc-0.10.1/doc/./data2hwdb", line 100, in main
lirc>     db = database.Database(configdir, yamldir)
lirc>   File "/build/lirc-0.10.1/python-pkg/lirc/database.py", line 135, in __init__
lirc>     cf = yaml.load(f.read())
lirc> TypeError: load() missing 1 required positional argument: 'Loader'
lirc> make[2]: *** [Makefile:1263: lirc.hwdb] Error 1
```

**Comment by dotlambda** [comment]
> Bisected another failure down to this bump on `staging`: `lirc`

We could apply https://sourceforge.net/p/lirc/git/merge-requests/39/

**Comment by trofi** [comment]
> > Bisected another failure down to this bump on `staging`: `lirc`
> 
> We could apply https://sourceforge.net/p/lirc/git/merge-requests/39/

Sounds good. Proposed possible fix as https://github.com/NixOS/nixpkgs/pull/150902

File: pkgs/development/python-modules/pyyaml/default.nix [hunk 1]
```diff
@@ -1,32 +1,32 @@
...
-, isPy27
 , python
 }:
 
 buildPythonPackage rec {
   pname = "PyYAML";
-  version = "5.4.1.1";
+  version = "6.0";
+
+  disabled = pythonOlder "3.6";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pyyaml/default.nix:14]: This breaks NixOps.
cc @NixOS/nixops-committers

**Comment by ulrikstrid** [comment]
This PR breaks 3.x of mongodb because it uses a older version of `scons` that is not compatible with python 3+ to build
```
nix-build -A mongodb-3_6              
error: PyYAML-6.0 not supported for interpreter python2.7
(use '--show-trace' to show detailed location information)
```

**Pull request merged**